chained/water_version_control.go

```diff
@@ -0,0 +1,136 @@
+package chained
+
+import (
+	"context"
+	"io"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+)
+
+type waterVersionControl struct {
+	dir string
+}
+
+type wasmInfo struct {
+	lastTimeLoaded time.Time
+	path           string
+}
+
+func newWaterVersionControl(dir string) *waterVersionControl {
+	return &waterVersionControl{
+		dir: dir,
+	}
+}
+
+// GetWASM returns the WASM file for the given transport
+// Remember to Close the io.ReadCloser after using it
+func (vc *waterVersionControl) GetWASM(ctx context.Context, transport string, downloader waterWASMDownloader) (io.ReadCloser, error) {
+	path := filepath.Join(vc.dir, transport+".wasm")
+	var f io.ReadCloser
+	f, err := os.Open(path)
+	if err != nil && !os.IsNotExist(err) {
+		return nil, log.Errorf("failed to open file %s: %w", path, err)
+	}
+
+	if os.IsNotExist(err) {
```

Comment:
From the godoc:

> This function predates [errors.Is](https://pkg.go.dev/errors#Is). It only supports errors returned by the os package. New code should use errors.Is(err, fs.ErrNotExist).

That said, I don't imagine this will create real problems here.  Just an FYI really.

Reply:
I'll follow the docs recommendations! Thanks for reminding!